core/src/Streamly/Internal/FileSystem/Path/Common.hs

```diff
@@ -463,3 +538,284 @@ append :: (Unbox a, Integral a) =>
+--
+unsafeSplitUNC :: (Unbox a, Integral a) => Array a -> (Array a, Array a)
+unsafeSplitUNC arr =
+    if cnt1 == 1 && unsafeIndexChar 2 arr == '?'
+    then do
+        if uncLen == 3
+                && unsafeIndexChar uncOff arr == 'U'
+                && unsafeIndexChar (uncOff + 1) arr == 'N'
+                && unsafeIndexChar (uncOff + 2) arr == 'C'
+        then unsafeSplitPrefix Windows (serverOff + serverLen) arr
+        else unsafeSplitPrefix Windows sepOff1 arr
+    else unsafeSplitPrefix Windows sepOff arr
+
+    where
+
+    len = Array.length arr
+    arr1 = Array.getSliceUnsafe 2 (len - 2) arr
+    cnt1 = countLeadingBy (not . isSeparatorWord Windows) arr1
+    sepOff = 2 + cnt1
+
+    -- XXX there should be only one separator in a valid path?
+    -- XXX it should either be UNC or two letter drive in a valid path
+    (uncOff, uncLen) = parseSegment arr len sepOff
+    sepOff1 = uncOff + uncLen
+    (serverOff, serverLen) = parseSegment arr len sepOff1
+
+-- XXX should we make the root Maybe? Both components will have to be Maybe to
+-- avoid an empty path.
+
+-- | If a path is rooted then separate the root and the remaining path
+-- otherwise root is returned as empty.
+--
+-- >>> toList (a,b) = (unpackPosix a, unpackPosix b)
+-- >>> splitPosix = toList . Common.splitRoot Common.Posix . packPosix
+--
+-- >>> splitPosix "/"
+-- ("/","")
+--
+-- >>> splitPosix "."
+-- (".","")
+--
+-- >>> splitPosix "/home"
+-- ("/","home")
+--
+-- >>> splitPosix "//"
+-- ("//","")
+--
+-- >>> splitPosix "./home"
+-- ("./","home")
+--
+-- >>> splitPosix "home"
+-- ("","home")
+--
+{-# INLINE splitRoot #-}
+splitRoot :: (Unbox a, Integral a) => OS -> Array a -> (Array a, Array a)
+splitRoot Posix arr
+    | isLocation Posix arr
+        = unsafeSplitTopLevel Posix arr
+    | otherwise = (Array.empty, arr)
+splitRoot Windows arr
+    | isRelativeCurDriveRoot arr || isRelativeCurDir Windows arr
+        = unsafeSplitTopLevel Windows arr
+    | hasDrive arr = unsafeSplitDrive arr
+    | isAbsoluteUNC arr = unsafeSplitUNC arr
+    | otherwise = (Array.empty, arr)
+
+-- | Split a path into components separated by the path separator. "."
+-- components in the path are ignored except in the leading position. Multiple
+-- consecutive separators are ignored.
+--
+-- >>> :{
+--  splitPosix = Stream.toList . fmap unpackPosix . Common.splitPath Common.Posix . packPosix
+--  splitWin = Stream.toList . fmap unpackWindows . Common.splitPath Common.Windows . packWindows
+-- :}
+--
+-- >>> splitPosix "."
+-- ["."]
+--
+-- >>> splitPosix "././"
+-- ["./"]
+--
+-- >>> splitPosix "./a/b/."
+-- ["./","a","b"]
+--
+-- >>> splitPosix "/"
+-- ["/"]
+--
+-- >>> splitPosix "/home"
+-- ["/","home"]
+--
+-- >>> splitWin "/home"
+-- ["/","home"]
+--
+-- >>> splitPosix "home//user/./..////\\directory/."
+-- ["home","user","..","\\directory"]
+--
+-- >>> splitWin "home//user/./..////\\directory/."
+-- ["home","user","..","directory"]
+--
+{-# INLINE splitPath #-}
+splitPath
```

Comment:
`splitPath` will ignore the trailing separator.
Is there a loss of information here?

Reply:
`a/b/` is the same as `a/b/.`
`a/b` is different from `a/b/`

Reply:
Ideally the components should not have separators. We can keep the trailing slash but that would require special casing it and not sure how useful that will be.